### PR TITLE
[JSON] Handle Search without ReJSON module

### PR DESCRIPTION
**Pull request by oshadmi**

### CODE REVIEW

File: src/json.c [hunk 1]
```diff
@@ -68,7 +68,7 @@ static RSLanguage SchemaRule_JsonScore(RedisModuleCtx *ctx, const SchemaRule *ru
     goto done;
   }
 
-  if(japi->getDoubleFromKey(jsonKey, rule->score_field, &score) != REDISMODULE_OK) {
+  if(japi && japi->getDoubleFromKey(jsonKey, rule->score_field, &score) != REDISMODULE_OK) {
```
Comment (rafie) [inline src/json.c:71]: If we're not supposed to even be here unless we have `japi` set, and we cannot return an error, we might as well just assert.

File: src/rules.c [hunk 2]
```diff
@@ -23,11 +24,11 @@ int SchemaRuleType_Parse(const char *type_str, SchemaRuleType *type, QueryError
   if (!type_str || !strcasecmp(type_str, RULE_TYPE_HASH)) {
     *type = SchemaRuleType_Hash;
     return REDISMODULE_OK;
-  } else if (!strcasecmp(type_str, RULE_TYPE_JSON)) {
+  } else if (japi && !strcasecmp(type_str, RULE_TYPE_JSON)) {
```
Comment (rafie) [inline src/rules.c:27]: In principle, we may allow creation of JSON indexes even if `japi` is not yet available, as it may become available if rejson is loaded after index creation. If no `japi` then no notifications, but also no operations to listen to, so we're sure not to miss anything.

File: src/rlookup.c [hunk 3]
```diff
@@ -359,12 +359,16 @@ static int getKeyCommonHash(const RLookupKey *kk, RLookupRow *dst, RLookupLoadOp
 
 static int getKeyCommonJSON(const RLookupKey *kk, RLookupRow *dst, RLookupLoadOptions *options,
                         RedisJSONKey *keyobj) {
+  if (!japi) {
+    QueryError_SetCode(options->status, QUERY_EUNSUPPTYPE);
```
Comment (rafie) [inline src/rlookup.c:363]: Maybe a nice little warning here.

File: src/json.c (hunk 1 above)
Comment (rafie) [reply src/json.c:71]: Don't we want to get out early here? `RedisJSON_GetString` is using `japi`.

File: src/rlookup.c [hunk 4]
```diff
@@ -595,7 +600,7 @@ static int RLookup_JSON_GetAll(RLookup *it, RLookupRow *dst, RLookupLoadOptions
   //RedisModuleString *krstr =
   //    RedisModule_CreateString(ctx, options->dmd->keyPtr, sdslen(options->dmd->keyPtr));
   // TODO: check error
-  RedisJSONKey jsonKey = japi->openKeyFromStr(ctx, options->dmd->keyPtr);
+  RedisJSONKey jsonKey = japi ? japi->openKeyFromStr(ctx, options->dmd->keyPtr) : NULL;
```
Comment (rafie) [inline src/rlookup.c:603]: I think it's better to explicitly ask for `!japi` here and to bail out. Going through `done` looks a little entangled to me.

Comment (oshadmi) [reply src/rlookup.c:603]: One more thing to consider is that this entangled situation is actually rather unexpected, and this is just additional precaution, leading to the "branch taken" approach (positive/sunny-day scenario).
Does this consideration make a difference?

File: src/json.c (hunk 1 above)
Comment (oshadmi) [reply src/json.c:71]: What line are you referring to?
Anyway, in `RedisJSON_GetString` there is also testing for none-null `japi`

**Comment by oshadmi** [comment]
@rafie Failures in tests (e.g. test_json:testAggregate, test_json:testAs, test_json:testDemo, test_json:testSearch in Ubuntu) are fixed by PR [Listen to REDISMODULE_NOTIFY_MODULE #1979](https://github.com/RediSearch/RediSearch/pull/1979)

**Review by rafie — APPROVED** [review]
